src/libNode/Node.cpp

```diff
@@ -246,265 +286,31 @@ bool Node::CheckState(Action action)
-                      "in WAITING_FINALBLOCK");
-            result = false;
-            break;
-        case ERROR:
-            LOG_GENERAL(WARNING,
-                        "Doing PROCESS_MICROBLOCKSUBMISSION but "
-                        "receiving ERROR message");
-            result = false;
-            break;
-        default:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Unrecognized or error state");
-            result = false;
-            break;
-        }
-        break;
-    case PROCESS_FINALBLOCK:
-        switch (m_state)
-        {
-        case POW1_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "POW1_SUBMISSION");
-            result = false;
-            break;
-        case POW2_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "POW2_SUBMISSION");
-            result = false;
-            break;
-        case TX_SUBMISSION:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in TX_SUBMISSION");
-            result = false;
-            break;
-        case TX_SUBMISSION_BUFFER:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "TX_SUBMISSION_BUFFER");
-            result = false;
-            break;
-        case MICROBLOCK_CONSENSUS_PREP:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "MICROBLOCK_CONSENSUS_PREP");
-            result = false;
-            break;
-        case MICROBLOCK_CONSENSUS:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Doing WAITING_FINALBLOCK but already in "
-                      "MICROBLOCK_CONSENSUS");
-            result = false;
-            break;
-        case WAITING_FINALBLOCK:
-            break;
-        case ERROR:
-            LOG_GENERAL(WARNING,
-                        "Doing WAITING_FINALBLOCK but receiving ERROR message");
-            result = false;
-            break;
-        default:
-            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-                      "Unrecognized or error state");
-            result = false;
-            break;
-        }
-        break;
-    default:
         LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
                   "Unrecognized action");
-        result = false;
-        break;
+        return false;
     }
 
-    return result;
+    if (m_state == ERROR)
```

Comment:
I would move this to the top, as a first check

Reply:
:D I've validated this and it's 100% same error messages with the original. If I move it it will print different error messages in some cases. I don't know if this change is important for people!

Reply:
Your changes made this extremely hard to read code much clearer :). The error checking is still weird, I guess you did it to preserve the existing order of checks.

IMO nobody should depend on error logs to make any decision and this makes more sense:

```
bool Node::CheckState(Action action)
{
    if (m_mediator.m_ds->m_mode != DirectoryService::Mode::IDLE)
    {
        LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
                                               string("I am a DS node.") +
                                               string(" Why am I getting this message? Action: ") << ActionString(action));
        return false;
    }

    if (m_state == ERROR)
    {
        LOG_GENERAL(WARNING, "Doing " << ActionString(action) << " but received ERROR message");
        return false;
    }

    if ((action == PROCESS_TXNBODY) || (action >= NUM_ACTIONS))
    {
        LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(), "Unrecognized action");
        return false;
    }

    if (action == m_state)
    {
        LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
                                                "Doing " << ActionString(action) << " but already in "
                                                << NodeStateString(m_state));
        return false;
    }

    // might want to log if it returns false to type, assuming not
    return compatibleState(m_state, action);
}
```

Reply:
Maybe you could confirm that what you posted has the same behavior as the original code. Specially the `action == m_state`.

> nobody should depend on error logs to make any decision
famous last words

In this commit, I do this transparent, zero-risk improvement. If people believe this isn't readable, they can do further commits in coordination with whoever operates the testnet. I don't want to change the logging. :)

Reply:
The thing is the original code should have never been written like and since it's not a real production system, now it's the time to do such changes. 

Maybe somebody who actually reads the logs can comment.

Reply:
OK, my comments is out of the scope of this pr.

Raised this https://github.com/Zilliqa/Zilliqa/issues/340